Add support for writingsuggestions attribute

rr-codes opened the pull request:
#### e1b69e3d4f52bbd35fe60e310459e2cc0ba1ba7a
<pre>
Add support for writingsuggestions attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=266824">https://bugs.webkit.org/show_bug.cgi?id=266824</a>
<a href="https://rdar.apple.com/114989563">rdar://114989563</a>

Reviewed by Aditya Keerthi.

Implement support for the new DOM `writingsuggstions` web API attribute, as defined by the
corresponding spec PR (whatwg/html#10018).

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isWritingSuggestionsEnabled const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/editing/VisibleSelection.cpp:
(WebCore::VisibleSelection::canEnableWritingSuggestions const):
* Source/WebCore/editing/VisibleSelection.h:
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::writingsuggestions const):
(WebCore::HTMLElement::setWritingsuggestions):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLElement.idl:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::supportsWritingSuggestions const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebKit/Shared/EditorState.cpp:
(WebKit::operator&lt;&lt;):
* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:
* Source/WebKit/Shared/FocusedElementInformation.h:
* Source/WebKit/Shared/FocusedElementInformation.serialization.in:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _updateTextInputTraits:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::postLayoutDataForContentEditable):
(WebKit::WebViewImpl::allowsInlinePredictions const):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getPlatformEditorStateCommon const):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::focusedElementInformation):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingSuggestions.mm: Added.
(-[WritingSuggestionsWebAPIWKWebView initWithHTMLString:]):
(-[WritingSuggestionsWebAPIWKWebView focusElementAndEnsureEditorStateUpdate:]):
(TEST):

Canonical link: <a href="https://commits.webkit.org/274912@main">https://commits.webkit.org/274912@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/157af428f0490ff249f99484746e05ad97e9605a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40273 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19285 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42651 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42818 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36362 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42580 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22215 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16614 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33447 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40847 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16241 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34754 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14021 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14067 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35707 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44096 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/33726 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36515 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36037 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39780 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/39899 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15088 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12363 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38061 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16707 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/46908 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9055 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16756 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9655 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16351 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingSuggestions.mm:
```diff
@@ -0,0 +1,117 @@
+TEST(WritingSuggestionsWebAPI, InlinePredictionType)
+{
+    {
+        auto webView = adoptNS([[WritingSuggestionsWebAPIWKWebView alloc] initWithHTMLString:@"<body><div id='div' contenteditable>This is some text.</div></body>"]);
```
Please make these separate tests.

annevk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4740,6 +4742,52 @@ bool Element::isSpellCheckingEnabled() const
+    // If none of the following conditions are true, then return `false`.
+
+    // `element` is an `input` element whose `type` attribute is in either the
+    // `Text`, `Search`, `URL`, `Email` state and is `mutable`.
```
You're not testing for mutable here and below: negation of `isDisabledFormControl()`.

annevk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4740,6 +4742,52 @@ bool Element::isSpellCheckingEnabled() const
+        if (!input)
+            return false;
+
+        auto isAllowedField = input->isTextField() || input->isSearchField() || input->isURLField() || input->isEmailField();
```
`isTextField()` encompasses all of these and more. I don't think it's what you want.

We probably want InputType's textTypes although that includes Telephone. Not including Telephone seems like a specification issue?

annevk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4740,6 +4742,52 @@ bool Element::isSpellCheckingEnabled() const
+    // `element` is an `input` element whose `type` attribute is in either the
+    // `Text`, `Search`, `URL`, `Email` state and is `mutable`.
+    auto isEligibleInputElement = [&] {
+        RefPtr input = dynamicDowncast<HTMLInputElement>(this);
```
*this to avoid null check?

annevk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4740,6 +4742,52 @@ bool Element::isSpellCheckingEnabled() const
+    // not in the `default` state and the nearest such ancestor's `writingsuggestions` content attribute
+    // is in the `false` state, then return `false`.
+
+    for (auto* ancestor = this; ancestor; ancestor = ancestor->parentOrShadowHostElement()) {
```
I'm pretty sure the specification was only gonna walk the normal tree. Did that change?

annevk commented on Source/WebCore/editing/VisibleSelection.cpp:
```diff
@@ -687,6 +687,20 @@ bool VisibleSelection::isInPasswordField() const
+    if (!containerNode)
+        return false;
+
+    RefPtr element = is<Element>(*containerNode) ? downcast<Element>(containerNode.get()) : containerNode->parentElement();
```
dynamicDowncast would be preferable even though this then becomes three lines.

annevk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4740,6 +4743,48 @@ bool Element::isSpellCheckingEnabled() const
+    }();
+
+    // `element` is a `textarea` element that is `mutable`.
+    auto isTextArea = is<HTMLTextAreaElement>(*this);
```
This also needs the disabled check. Not sure about the editing host stuff below.

annevk commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -1043,6 +1043,18 @@ bool HTMLInputElement::isTextType() const
     return m_inputType->isTextType();
 }
 
+bool HTMLInputElement::supportsWritingSuggestions() const
```
We really need to refactor how this works. I get the feeling we're being quite inconsistent about this across the code base. (Not a blocker for this patch though.)

whsieh commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4740,6 +4743,48 @@ bool Element::isSpellCheckingEnabled() const
+    auto isTextArea = is<HTMLTextAreaElement>(*this);
+
+    // `element` is an `editing host` or is `editable`.
+    auto isEditable = hasEditableStyle() || isContentEditable();
```
It's unclear why we need to check both bits here. I would recommend just checking `hasEditableStyle()`, and moving this flag into `EditorState::PostLayoutData`, along with the other bits of information on `EditorState` that already rely on layout being up-to-date.

rr-codes replied:
> moving this flag
which flag is this referring to? this entire method `isWritingSuggestionsEnabled`?

pxlcoder replied:
It's referring to the flag you have on `EditorState`.

pxlcoder commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4753,6 +4756,54 @@ bool Element::isSpellCheckingEnabled() const
+            return false;
+
+        return !textArea->isDisabledFormControl();
+    }();
```
Isn't it better to call the lambdas in the `if`, for lazy evaluation. As it is there's some redundant checking going on.

pxlcoder commented on Source/WebCore/editing/VisibleSelection.cpp:
```diff
@@ -687,6 +687,22 @@ bool VisibleSelection::isInPasswordField() const
+bool VisibleSelection::canEnableWritingSuggestions() const
+{
+    RefPtr containerNode = start().containerNode();
+
```
I would remove this newline.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -531,6 +531,8 @@
 
     auto& postLayoutData = *result.postLayoutData;
 
+    postLayoutData.canEnableWritingSuggestions = selection.canEnableWritingSuggestions();
```
Why not put this under the `result.isContentEditable` below?

Should be okay because `result.isContentEditable = selection.hasEditableStyle();`.

rr-codes committed:
Add support for writingsuggestions attribute
https://bugs.webkit.org/show_bug.cgi?id=266824
rdar://114989563

Reviewed by Aditya Keerthi.

Implement support for the new DOM `writingsuggstions` web API attribute, as defined by the
corresponding spec PR (whatwg/html#10018).

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isWritingSuggestionsEnabled const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/editing/VisibleSelection.cpp:
(WebCore::VisibleSelection::canEnableWritingSuggestions const):
* Source/WebCore/editing/VisibleSelection.h:
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::writingsuggestions const):
(WebCore::HTMLElement::setWritingsuggestions):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLElement.idl:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::supportsWritingSuggestions const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebKit/Shared/EditorState.cpp:
(WebKit::operator<<):
* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:
* Source/WebKit/Shared/FocusedElementInformation.h:
* Source/WebKit/Shared/FocusedElementInformation.serialization.in:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _updateTextInputTraits:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::postLayoutDataForContentEditable):
(WebKit::WebViewImpl::allowsInlinePredictions const):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getPlatformEditorStateCommon const):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::focusedElementInformation):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingSuggestions.mm: Added.
(-[WritingSuggestionsWebAPIWKWebView initWithHTMLString:]):
(-[WritingSuggestionsWebAPIWKWebView focusElementAndEnsureEditorStateUpdate:]):
(TEST):

Canonical link: https://commits.webkit.org/274912@main
